### PR TITLE
Replace shell script with python zimit.py + crawl dedup improvements

### PR DESCRIPTION
**Pull request by ikreymer**
Using python zimit.py for arg parsing + warc2zim calls as suggested in #28. Should fix:
- #18, #28 - arg parsing now handled in python, no shell
- #25 - better handling of interrupt signals, should shutdown gracefully.
- crawler.js now only handles crawling (may be moved to separate project later).
- remove zimit user, su, and run all as root user (needing to add --no-sandbox for chrome)

Also includes crawling update to improve cacheing/reduce disk usage:
- Using latest version of pywb, which deduplicates all URLs (using Redis), so duplicates are never stored in WARCs, should reduce disk space.
- Forces browser cacheing on existing urls to reduce load

### CODE REVIEW

**Comment by ikreymer** [comment]
I don't know how to disable the partial path CodeFactor errors (tried with config file), they don't apply here since this is designed to run in a container with known executables.

**Review by rgaudin — CHANGES_REQUESTED** [review]
Some questions, couple issues. I've fixed the partial paths thing.

File: Dockerfile [hunk 1]
```diff
@@ -13,31 +13,26 @@ ENV PROXY_HOST=localhost \
     PROXY_CA_FILE=/tmp/proxy-ca.pem \
     NO_SOCAT=1
 
-RUN pip install gevent>=20.9.0 pywb uwsgi
+RUN pip install gevent>=20.9.0 uwsgi
 
 RUN pip install warc2zim==1.1.0
 
+RUN pip install git+https://github.com/webrecorder/pywb@patch-work
```
[the remark this answers is not in the document]
Comment (ikreymer) [reply Dockerfile:20]: Yes, I can make it be `@develop` but its not quite ready for release there.., can update when its out in the next version of pywb.

[the remark this answers is not in the document]
Comment (rgaudin) [reply Dockerfile:20]: No it's ok, just wanted to be sure

File: crawler.js [hunk 2]
```diff
@@ -13,12 +13,23 @@ const HTTPS_AGENT = require("https").Agent({
...
+});
+
 
 async function run(params) {
   // Chrome Flags, including proxy server
   const args = [
     "--no-xshm", // needed for Chrome >80 (check if puppeteer adds automatically)
-    `--proxy-server=http://${process.env.PROXY_HOST}:${process.env.PROXY_PORT}`
+    `--proxy-server=http://${process.env.PROXY_HOST}:${process.env.PROXY_PORT}`,
+    "--no-sandbox"
```
[the remark this answers is not in the document]
Comment (ikreymer) [reply crawler.js:32]: This allows running chrome as root user. It is not necessarily ideal in general, but I think fine in this scenario, and given the constraints of only command-line options. The other main solution is to run with a custom security profile, which you'd need to store on zimfarm outside docker, eg:
https://github.com/armbues/chrome-headless/blob/master/README.md

But maybe worth reexamining once other issues are fixed.

File: zimit.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+from warc2zim.main import warc2zim
+
+
+def zimit(args=None):
+    parser = ArgumentParser(
+        description="Run a browser-based crawl on the specified URL and convert to ZIM"
+    )
+
+    parser.add_argument("-u", "--url", help="The URL to start crawling from")
```
[the remark this answers is not in the document]
Comment (ikreymer) [reply zimit.py:29]: Ok, yeah, probably it should be handled in warc2zim itself.. I'll add a quick PR there.

File: zimit.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+    parser.add_argument(
+        "--limit", help="Limit crawl to this number of pages", type=int, default=0
+    )
+
+    parser.add_argument(
+        "--timeout",
+        help="Timeout for each page to load (in millis)",
+        type=int,
+        default=90000,
```
[the remark this answers is not in the document]
Comment (ikreymer) [reply zimit.py:47]: Sure

File: zimit.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+    )
+
+    parser.add_argument(
+        "--scope",
+        help="The scope of current page that should be included in the crawl (defaults to the domain of URL)",
```
[the remark this answers is not in the document]
Comment (ikreymer) [reply zimit.py:52]: It's actually the other way now. I think from crawling perspective, it makes more sense that if you have `--url http://domdom.tld/category1/start.html`, the scope is`http://domdom.tld/category1/`.
If you want it to be the whole domain you'd add `--scope http://domdom.tld`. This seemed like a better default, but can change it the other way if you'd prefer that..

One use case for this is for example `https://journals.openedition.org/bibnum/`, where you don't expect to crawl all of `https://journals.openedition.org/` when you specify `--url https://journals.openedition.org/bibnum/`

I think its good to err on the side of more conservative crawling scope, and specify a more broader one explicitly.

File: zimit.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+    )
+
+    parser.add_argument(
```
[the remark this answers is not in the document]
Comment (ikreymer) [reply zimit.py:55]: Only pages

File: zimit.py [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+
+    cmd_args = get_node_cmd_line(zimit_args)
+    cmd_line = " ".join(cmd_args)
+
+    print("")
+    print("----------")
+    print("running zimit driver: " + cmd_line)
+    subprocess.run(cmd_args, check=True)  # nosec
+
+    warc_files = glob.glob(
```
[the remark this answers is not in the document]
Comment (ikreymer) [reply zimit.py:135]: I don't think the list will be big enough to cause any issues, even if hundreds of warcs.
This list is also passed to warc2zim for iteration. We could support passing a directory of warcs in warc2zim instead then. Do you prefer that?

File: zimit.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (rgaudin) [reply zimit.py:52]: Agrees. Help text should be adjusted then.

File: zimit.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (rgaudin) [reply zimit.py:135]: yeah I think it'd make more sense actually.

File: zimit.py [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+        os.path.join(temp_root_dir, "collections/capture/archive/*.warc.gz")
+    )
+    print("")
+    print("----------")
+    print("Processing {0} WARC files to ZIM".format(len(warc_files)))
+
+    res = warc2zim(warc2zim_args + warc_files)
```
[the remark this answers is not in the document]
Comment (ikreymer) [reply zimit.py:142]: i think was checking it before, can change.

File: zimit.py [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+    print("----------")
+    print("Testing warc2zim args")
+    print("Running: warc2zim " + " ".join(warc2zim_args))
+    res = warc2zim(warc2zim_args)
+    if res != 100:
+        print("Exiting, invalid warc2zim params")
+        return 2
+
+    # make temp dir for this crawl and make it all writeable+all readable+all exec
+    temp_root_dir = tempfile.mkdtemp(dir=zimit_args.output, prefix=".tmp")
```
[the remark this answers is not in the document]
Comment (ikreymer) [reply zimit.py:96]: Hm, pathlib doesn't have a mkdir (and the comment is out of date, no longer adjusting permissions will fix)

[the remark this answers is not in the document]
Comment (rgaudin) [reply zimit.py:96]: I meant wrapping it in a Path() as mkdtemp returns a string and we try to always use pathlib for paths

File: zimit.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (rgaudin) [reply zimit.py:142]: ok

File: zimit.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ikreymer) [reply zimit.py:30]: Handled in warc2zim

**Review by rgaudin — APPROVED** [review]

**Pull request merged**